lib/variable/reduction.cpp

```diff
@@ -22,148 +22,104 @@ using namespace scipp::core;
 namespace scipp::variable {
 
 namespace {
-
-Variable make_accumulant(const Variable &var, const Dim dim,
+/// Note that masking is not supported here since it would make creation
+/// of a sensible starting value difficult. So masks must be applied
+/// before calling reduce_idempotent.
+Variable apply_reduction(const Variable &var, const Dim dim,
+                         void (&op)(Variable &, const Variable &),
                          const FillValue &init) {
-  if (variableFactory().has_masks(var))
-    throw except::NotImplementedError(
-        "Reduction operations for binned data with "
-        "event masks not supported yet.");
   auto dims = var.dims();
   dims.erase(dim);
-  auto prototype = empty(dims, variableFactory().elem_unit(var),
-                         variableFactory().elem_dtype(var),
-                         variableFactory().has_variances(var));
-  return special_like(prototype, init);
+  auto reduced = make_reduction_accumulant(var, dims, init);
+  reduce_into(reduced, var, op);
+  return reduced;
```

Comment:
This is just `reduce_bins` from `dataset/bins_reduction.cpp`. Can we move this here and call it (and rename it?)?

Reply:
Not quite. `reduce_bins` determines `dims`. I didn't want to combine the two because in non-bin reductions, not passing `dim` means that all dimensions are reduced. But for bins it means that the event dimensions are reduced. Is there an API that avoids this confusion?

Reply:
Right, you mean it used `var.dims()`? Can we just turn the above 3 lines into a function and use that for the bins reduction (passing `var.dims()`)? Then it will be clear that this is doing the same (it took me a while to figure out that it is not calling some other function of the same name, etc.).

Reply:
Is that still a problem with the latest changes?

Reply:
Solved as far as I can tell.